services/cronjob/dataproxyrefresher/cmd/dataproxyrefresher/main.go

```diff
@@ -25,231 +26,421 @@ func init() {
+}
+
+func (r *DataproxyRefresher) validateProfile(
+	profileJSON map[string]interface{},
+) (bool, string, error) {
+	validateURL := config.Conf.Index.URL + APIValidatePath
+
+	isValid, failureReasons, err := importutil.Validate(
+		validateURL,
+		profileJSON,
+	)
+	if err != nil {
+		return false, "",
+			fmt.Errorf("error during profile validation: %w", err)
+	}
+
+	return isValid, failureReasons, nil
+}
+
+func (r *DataproxyRefresher) UpdateProfile(
+	profile model.Profile,
+	profileJSON map[string]interface{},
+) error {
+	// Check if the profile already exists.
+	count, err := r.svc.Count(profile.Oid)
+	if err != nil {
+		return fmt.Errorf(
+			"can't count profile. Profile ID: %s, error: %v",
+			profile.Oid,
+			err,
+		)
+	}
+
+	if count <= 0 {
+		// If profile doesn't exist, create a new one.
+		profileJSON["cuid"] = cuid.New()
+		if err := r.svc.Add(profileJSON); err != nil {
+			return fmt.Errorf(
+				"can't add a profile. Profile ID: %s, error: %v",
+				profile.Oid,
 				err,
 			)
-			cleanUp()
 		}
-
-		var profileData []interface{}
-		err = json.Unmarshal(bodyBytes, &profileData)
+	} else {
+		// If profile already exists, update it.
+		result, err := r.svc.Update(profile.Oid, profileJSON)
 		if err != nil {
-			logger.Error(
-				"Failed to unmarshal data from API. Profile CUID:"+profile.Cuid,
-				err,
+			return fmt.Errorf(
+				"can't update a profile. Profile ID: %s, error: %v",
+				profile.Oid, err,
 			)
-			cleanUp()
 		}
+		profileJSON["cuid"] = result["cuid"]
+	}
 
-		// If the node still exist, don't delete it and update access_time
-		if len(profileData) > 0 {
-			profileJSON := importutil.MapFieldsName(
-				profileData[0].(map[string]interface{}),
-				mapping,
-			)
-			doc, err := json.Marshal(profileJSON)
-			if err != nil {
-				logger.Error(
-					"Failed to marshal data. Profile CUID: "+profile.Cuid,
-					err,
-				)
-				cleanUp()
-			}
+	// Post the update to the Index
+	postNodeURL := config.Conf.Index.URL + APINodesPath
+	profileURL := config.Conf.DataProxy.URL + APIProfilesPath + "/" +
+		profileJSON["cuid"].(string)
+	nodeID, err := importutil.PostIndex(postNodeURL, profileURL)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to post profile to Index. Profile URL: %s, error: %v",
+			profileURL,
+			err,
+		)
+	}
 
-			// TODO
-			profileHash, err := jsonutil.Hash(string(doc))
-			if err != nil {
-				logger.Error(
-					"Failed to hash data. Profile CUID: "+profile.Cuid,
-					err,
-				)
-				cleanUp()
-			}
+	// Save node_id to profile
+	if err := r.svc.UpdateNodeID(profile.Oid, nodeID); err != nil {
+		return fmt.Errorf(
+			"update node id failed. Profile ID: %s, error: %v",
+			profile.Oid,
+			err,
+		)
+	}
 
-			if profileHash != profile.SourceDataHash {
-				logger.Info(
-					"Source data hash mismatch: " + profile.Cuid + " - " + profile.Oid + " : " + profile.SourceDataHash + " - " + profileHash,
-				)
-
-				// reconstruct data
-				profileJSON, err = importutil.MapProfile(
```

Comment:
I believe these deleted lines are crucial component.
The profileJSON we generated in the past only mapped the fields and values.
Right before updating the profile in the Index, it's imperative that we incorporate `importutil.MapProfile` to include additional information for converting KVM to organizational schemas.

Reply:
That's right. We need to keep the profile mapping here.

Reply:
@TingSyuanWang I have updated it. Please have check.